### PR TITLE
Add Poisson/Helmholtz test: MatrixFree + simplex mesh

### PR DESCRIPTION
**Pull request by peterrum**
References #11166.

~Depends on #11186 and on #11188.~

### CODE REVIEW

File: include/deal.II/matrix_free/mapping_info.templates.h [hunk 1]
```diff
@@ -369,12 +410,20 @@ namespace internal
...
+
+              if (flag == false)
                 {
-                  Assert(quad[my_q][hpq].get_tensor_basis()[0] ==
-                           quad[my_q][hpq].get_tensor_basis()[i],
-                         ExcNotImplemented());
+#ifdef DEAL_II_WITH_SIMPLEX_SUPPORT
+                  continue; // TODO: face data is not set up
+#else
+                  Assert(false, ExcNotImplemented())
```
Comment (drwells) [inline include/deal.II/matrix_free/mapping_info.templates.h:425]:
```suggestion
                  Assert(false, ExcNotImplemented());
```

File: include/deal.II/matrix_free/mapping_info.templates.h [hunk 2]
```diff
@@ -338,27 +365,41 @@ namespace internal
...
```
Comment (drwells) [inline include/deal.II/matrix_free/mapping_info.templates.h:401]:
```suggestion
                  Assert(false, ExcNotImplemented());
```

File: tests/simplex/matrix_free_02.cc [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  SolverCG<LinearAlgebra::distributed::Vector<double>> solver(
+    reduction_control);
+  solver.solve(poisson_operator, x, b, PreconditionIdentity());
+
+  if (Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0)
+    printf("Solved in %d iterations.\n", reduction_control.last_step());
+
+  constraints.distribute(x);
+
+#if false
```
Comment (drwells) [inline tests/simplex/matrix_free_02.cc:169]:
```suggestion
#if 0
```

I suspect (but I don't have evidence on hand) that some toolchains might print a warning with `#if false` so I'd prefer the unambiguously false version.

**Review by drwells — DISMISSED** [review]
I have some minor requests.

Why don't things work with 3D? It's fine if its not in the PR but I don't see why this is specific to 2D.

**Comment by peterrum** [comment]
> Why don't things work with 3D? It's fine if its not in the PR but I don't see why this is specific to 2D.

I can do that. Probably, it would be also good to compare with an actual sparse matrix vector multiplication. I'll check how the tests are structured in the `matrix_free` test folder.

File: tests/simplex/matrix_free_02.cc [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    matrix_free.template cell_loop<VectorType, VectorType>(
+      [&](const auto &, auto &dst, const auto &src, const auto cells) {
+        FEEvaluation<dim, -1, 0, 1, double> phi(matrix_free);
+        for (unsigned int cell = cells.first; cell < cells.second; ++cell)
+          {
+            phi.reinit(cell);
+            phi.gather_evaluate(src, false, true);
+
+            for (unsigned int q = 0; q < phi.n_q_points; ++q)
+              phi.submit_gradient(phi.get_gradient(q), q);
```
Comment (kronbichler) [inline tests/simplex/matrix_free_02.cc:98]: Do we need a test that also goes through the `evaluate(EvaluationFlags::values)` case? If I understand things correctly that part is not yet tested (whereas the integrate part is due to the rhs). Nothing wrong with testing the Laplacian, but I just want to explain why the Helmholtz operator is my favorite: https://github.com/dealii/dealii/blob/7b87282453f779f47413a126b448c3fec7e73c2d/tests/matrix_free/matrix_vector_mf.h#L47-L51

**Comment by peterrum** [comment]
@drwells @kronbichler I have extended the test so that it tests a Poisson problem and a Helmholtz problem for 2D and 3D, as well as, for linear and quadratic elements. The results are compared to a matrix-based implementation.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**